src/3d/qgsrubberband3d.cpp

```diff
@@ -38,51 +38,46 @@
 #include <Qt3DCore/QGeometry>
 #endif
 
+#include "qgs3dutils.h"
+#include "qgsmessagelog.h"
+#include "qgstessellatedpolygongeometry.h"
+
 #include <Qt3DRender/QGeometryRenderer>
 #include <QColor>
 
 
 /// @cond PRIVATE
 
 
-QgsRubberBand3D::QgsRubberBand3D( Qgs3DMapSettings &map, QgsWindow3DEngine *engine, Qt3DCore::QEntity *parentEntity, Qgis::GeometryType geometryType )
+QgsRubberBand3D::QgsRubberBand3D( Qgs3DMapSettings &map, QgsWindow3DEngine *engine, Qt3DCore::QEntity *parentEntity, const Qgis::GeometryType geometryType, const bool isOutlineDashed )
   : mMapSettings( &map )
   , mEngine( engine )
   , mGeometryType( geometryType )
+  , mIsOutlineDashed( isOutlineDashed )
 {
-  if ( mGeometryType == Qgis::GeometryType::Line || mGeometryType == Qgis::GeometryType::Polygon )
+  switch ( mGeometryType )
   {
-    // Rubberband line
-    mLineEntity = new Qt3DCore::QEntity( parentEntity );
-
-    QgsLineVertexData dummyLineData;
-    mGeometry = dummyLineData.createGeometry( mLineEntity );
-
-    Q_ASSERT( mGeometry->attributes().count() == 2 );
-    mPositionAttribute = mGeometry->attributes().at( 0 );
-    mIndexAttribute = mGeometry->attributes().at( 1 );
-
-    mLineGeomRenderer = new Qt3DRender::QGeometryRenderer;
-    mLineGeomRenderer->setPrimitiveType( Qt3DRender::QGeometryRenderer::LineStripAdjacency );
-    mLineGeomRenderer->setGeometry( mGeometry );
-    mLineGeomRenderer->setPrimitiveRestartEnabled( true );
-    mLineGeomRenderer->setRestartIndexValue( 0 );
-
-    mLineEntity->addComponent( mLineGeomRenderer );
-
-    mLineMaterial = new QgsLineMaterial;
-    mLineMaterial->setLineWidth( mWidth );
-    mLineMaterial->setLineColor( mColor );
-
-    QObject::connect( engine, &QgsAbstract3DEngine::sizeChanged, mLineMaterial, [this, engine] {
-      mLineMaterial->setViewportSize( engine->size() );
-    } );
-    mLineMaterial->setViewportSize( engine->size() );
-
-    mLineEntity->addComponent( mLineMaterial );
+    case Qgis::GeometryType::Point:
+      setupMarker( parentEntity );
+      break;
+    case Qgis::GeometryType::Line:
+      setupLine( parentEntity, engine );
+      setupMarker( parentEntity );
+      break;
+    case Qgis::GeometryType::Polygon:
+      setupMarker( parentEntity );
+      setupLine( parentEntity, engine );
+      setupPolygon( parentEntity );
+      break;
+    case Qgis::GeometryType::Null:
+    case Qgis::GeometryType::Unknown:
+      QgsMessageLog::logMessage( QObject::tr( "Unknown GeometryType used in QgsRubberband3D" ), QObject::tr( "3D" ) );
```

Comment:
I would just use `QgsDebugError` and no tr() - this is a programming error rather than something that user needs to see (and thus be translatable)